ddl/ddl.go

```diff
@@ -488,10 +488,22 @@ func (d *ddl) doDDLJob(ctx sessionctx.Context, job *model.Job) error {
 		metrics.HandleJobHistogram.WithLabelValues(job.Type.String(), metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
 	}()
 	for {
+		failpoint.Inject("storeCloseInLoop", func(_ failpoint.Value) {
+			d.cancel()
+		})
+
+		// avoid data race with restartWorkers in ddl_test.go
+		d.mu.RLock()
```

Comment:
Could we add this code to [d.doDDLJob(ctx, job)](https://github.com/pingcap/tidb/pull/18844/files#diff-4226364954d30f0f8c2648cb19da228eR239)? Then we needn't change this code.

Reply:
i guess if we protect whole function, restartWorker can't run concurrently with doDDLJob, so there's not a test for "interrupted job"

Reply:
done in https://github.com/pingcap/tidb/pull/18844/commits/f7eb7d36d09608b4460fa849fd65023a8992c7f4

Reply:
> guess if we protect whole function, restartWorker can't run concurrently with doDDLJob, so there's not a test for "interrupted job"

Yes. But we'd better don't change original code because of the test.